graph/path/bellman_ford_moore.go

```diff
@@ -40,33 +96,46 @@ func BellmanFordFrom(u graph.Node, g graph.Graph) (path Shortest, ok bool) {
 				if !ok {
 					panic("bellman-ford: unexpected invalid weight")
 				}
-				joint := path.dist[j] + w
-				if joint < path.dist[k] {
-					path.set(k, joint, j)
-					changed = true
+				if path.dist[j]+w < path.dist[k] {
+					path.hasNegativeCycle = true
+					return path, false
 				}
 			}
 		}
-		if !changed {
-			break
-		}
 	}
 
-	for j, u := range nodes {
-		uid := u.ID()
-		for _, v := range graph.NodesOf(g.From(uid)) {
-			vid := v.ID()
-			k := path.indexOf[vid]
-			w, ok := weight(uid, vid)
-			if !ok {
-				panic("bellman-ford: unexpected invalid weight")
-			}
-			if path.dist[j]+w < path.dist[k] {
-				path.hasNegativeCycle = true
-				return path, false
-			}
-		}
+	return path, true
+}
+
+// bellmanFordQueue is a queue for the Queue-based Bellman ford algorithm
+type bellmanFordQueue struct {
+	nodes []graph.Node
+}
+
+// enqueue adds a node to the bellmanFordQueue
+func (b *bellmanFordQueue) enqueue(n graph.Node) {
+	b.nodes = append(b.nodes, n)
+}
+
+// dequeue returns the first value of the bellmanFordQueue
+func (b *bellmanFordQueue) dequeue() (graph.Node, error) {
+	if len(b.nodes) == 0 {
+		return nil, errors.New("queue is empty!")
```

Comment:
This error is not necessary since if we got here is was due to a programmer error and should be a panic (the `linear.NodeQueue` does this.